ngraph/core/include/ngraph/op/constant.hpp

```diff
@@ -335,8 +363,52 @@ namespace ngraph
                     m_alloc_buffer_on_visit_attributes = val;
                 }
 
-            protected:
-                template <element::Type_t Type, typename OUT_T>
+            private:
+                template <element::Type_t Type,
+                          typename DataStorageType = fundamental_type_for<Type>,
+                          typename std::enable_if<Type != element::Type_t::u1 &&
+                                                      Type != element::Type_t::u4 &&
+                                                      Type != element::Type_t::i4,
+                                                  bool>::type = true>
+                DataStorageType get_element_value(size_t index) const
+                {
+                    return get_data_ptr<Type>()[index];
+                }
+
+                template <element::Type_t Type,
+                          typename DataStorageType = fundamental_type_for<Type>,
+                          typename std::enable_if<Type == element::Type_t::u1, bool>::type = true>
+                DataStorageType get_element_value(size_t index) const
+                {
+                    return (get_data_ptr<uint8_t>()[index / 8] >> (7 - (index % 8))) & 1;
+                }
+
+                template <element::Type_t Type,
+                          typename DataStorageType = fundamental_type_for<Type>,
+                          typename std::enable_if<Type == element::Type_t::u4, bool>::type = true>
+                DataStorageType get_element_value(size_t index) const
+                {
+                    return (get_data_ptr<uint8_t>()[index / 2] >> (index % 2 ? 0 : 4)) & 0x0F;
+                }
+
+                template <element::Type_t Type,
+                          typename DataStorageType = fundamental_type_for<Type>,
+                          typename std::enable_if<Type == element::Type_t::i4, bool>::type = true>
+                DataStorageType get_element_value(size_t index) const
+                {
+                    const uint8_t i4data =
+                        (get_data_ptr<uint8_t>()[index / 2] >> (index % 2 ? 0 : 4)) & 0x0F;
+                    const bool is_negative_number = (i4data >> 3) & 0b1;
+                    const int8_t data = is_negative_number ? i4data | 0xF0 : i4data;
+                    return data;
+                }
+
+                template <element::Type_t Type,
+                          typename OUT_T,
+                          typename std::enable_if<Type != element::Type_t::u1 &&
+                                                      Type != element::Type_t::u4 &&
+                                                      Type != element::Type_t::i4,
+                                                  bool>::type = true>
                 void cast_vector(std::vector<OUT_T>& output_vector) const
```

Comment:
Since you're already refactoring this class, would it be a lot of work to change this method's name to something more aligned with what it actually does? It doesn't cast any vector but rather casts the underlying buffer data to a vector. Even `cast_to_vector` would have been better.

Reply:
We cannot change the public API. In other case we need to introduce the new version of Constant operation.
This PR extends the set of supported types but doesn't change the operation API

Reply:
Actually, this method was `protected` and now is `private`. So should I rename this function?

Reply:
In master it is public https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/op/constant.hpp#L271 so I guess we'll have to leave it as is.

Reply:
In master [std::vector<T> cast_vector() const](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/op/constant.hpp#L271) is public and [void cast_vector(std::vector<OUT_T>& output_vector) const] (https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/op/constant.hpp#L340) is protected.
Protected implementation is change to private in this PR.

Reply:
Ok, got it now. I guess there's no point in refactoring only the private part if the public api can't be changed.